File: frontend/src/domain/intentResult.ts
import { z } from "zod";

export enum IntentRejectionCode {
  WrongPhase = 1,
  NotAuthorized = 2,
  UnknownSession = 3,
  InvariantViolated = 4,
  MalformedPayload = 5,
  UnknownParticipant = 6,
}

export const intentResultSchema = z.object({
  isAccepted: z.boolean(),
  code: z.enum(IntentRejectionCode).nullable(),
  detail: z.string().nullable(),
});

export type IntentResult = z.infer<typeof intentResultSchema>;

File: backend/Application/Intents/IntentResult.cs
namespace ValuesWorkshop.Application.Intents;

public sealed record IntentResult(bool IsAccepted, IntentRejectionCode? Code, string? Detail)
{
    public static IntentResult Accepted()
    {
        return new IntentResult(true, null, null);
    }

    public static IntentResult Rejected(IntentRejectionCode code, string detail)
    {
        return new IntentResult(false, code, detail);
    }
}
